components/brave_vpn/brave_vpn_connection_manager_mac.mm

```diff
@@ -0,0 +1,266 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_vpn/brave_vpn_connection_manager_mac.h"
+
+#import <Foundation/Foundation.h>
+
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/mac/bundle_locations.h"
+#include "base/mac/foundation_util.h"
+#include "base/strings/sys_string_conversions.h"
+
+// Referenced GuardianConnect implementation.
+// https://github.com/GuardianFirewall/GuardianConnect
+namespace brave_vpn {
+
+namespace {
+
+const NSString* kBraveVPNKey = @"BraveVPNKey";
+
+NSData* GetPasswordRefForAccount(const NSString* account_key) {
+  if (account_key == nil || [account_key length] == 0) {
+    LOG(ERROR) << "Error: account key is empty";
+    return nil;
+  }
+
+  NSString* bundle_id = [[NSBundle mainBundle] bundleIdentifier];
+  CFTypeRef copy_result = NULL;
+  NSDictionary* query = @{
+    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
+    (__bridge id)kSecAttrService : bundle_id,
+    (__bridge id)kSecAttrAccount : account_key,
+    (__bridge id)kSecMatchLimit : (__bridge id)kSecMatchLimitOne,
+    (__bridge id)kSecReturnPersistentRef : (__bridge id)kCFBooleanTrue,
+  };
+  OSStatus results =
+      SecItemCopyMatching((__bridge CFDictionaryRef)query, &copy_result);
+  if (results != errSecSuccess)
+    LOG(ERROR) << "Error: obtaining password ref(status:" << results << ")";
+  return (__bridge NSData*)copy_result;
+}
+
+OSStatus RemoveKeychainItemForAccount(const NSString* account_key) {
+  if (account_key == nil || [account_key length] == 0) {
+    LOG(ERROR) << "Error: account key is empty";
+    return errSecParam;
+  }
+
+  NSString* bundle_id = [[NSBundle mainBundle] bundleIdentifier];
+  NSDictionary* query = @{
+    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
+    (__bridge id)kSecAttrService : bundle_id,
+    (__bridge id)kSecAttrAccount : account_key,
+    (__bridge id)kSecReturnPersistentRef : (__bridge id)kCFBooleanTrue,
+  };
+  OSStatus result = SecItemDelete((__bridge CFDictionaryRef)query);
+  if (result != errSecSuccess && result != errSecItemNotFound)
+    LOG(ERROR) << "Error: deleting password entry(status:" << result << ")";
+  return result;
+}
+
+OSStatus StorePassword(const NSString* password, const NSString* account_key) {
+  if (password == nil || [password length] == 0) {
```

Comment:
Do we need to validate password input sizes here and/or enforce a minimum besides zero?

Reply:
Yes, checking with specific length range would be helpful.
I'll check if we have a spec for the password length. If not now, I'll left a TODO comment to f/u in the future.

Reply:
@bsclifton if we can know password spec such as minimum length, it can be used here for validation.
but I assume we don't know now?

Reply:
@simonhong good question - for now, I don't think we know. But if we do find out, would be great to update